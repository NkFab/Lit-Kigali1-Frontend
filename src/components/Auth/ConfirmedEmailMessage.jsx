import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { fetchConfirmEmail } from '../../redux/actions/signupActions';

export class ConfirmedEmailMessage extends Component {
  state = {
    error: false,
    message: '',
    loading: true,
  };

  componentDidMount() {
    const {
      onEmailVerification,
      match: { params },
    } = this.props;
    onEmailVerification({ ...params });
  }

  render() {
    const { confirmMessage } = this.props;

    return (
      <section className="main-content content-margin">
        <div className="container align-center">
          <p className="color-primary title-2 align-center">{confirmMessage}</p>
        </div>
      </section>
    );
  }
}

ConfirmedEmailMessage.propTypes = {
  confirmMessage: PropTypes.string,
  onEmailVerification: PropTypes.func.isRequired,
};
ConfirmedEmailMessage.defaultProps = {
  confirmMessage: '',
};
export const mapStateToProps = state => ({
  confirmMessage: state.signUp.confirmMessage,
});

export const mapDispatchToProps = dispatch => ({
  onEmailVerification: ({ userId, confirmationCode }) => dispatch(fetchConfirmEmail({ userId, confirmationCode })),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ConfirmedEmailMessage);
