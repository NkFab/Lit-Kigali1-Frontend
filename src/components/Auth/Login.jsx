import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import fb from '../../assets/images/facebook.png';
import twitter from '../../assets/images/twitter.png';
import google from '../../assets/images/google.png';
import Button from '../common/Button/Button';
import Input from '../common/Input/Input';
import { loginUser, inputHandler, validateCredentials } from '../../redux/actions';


class LoginComponent extends Component {
  signin = (e) => {
    const { username, password } = this.props.credentials;
      this.props.validateCredentials({ username, password })
      .then((res) => {
        if(res.message === 'Ok') {
          this.props.loginUser({ username, password });
        }
      });
    e.preventDefault();
  };

  handleInput = (e) => {
    this.props.inputHandler({ field: e.target.name, value: e.target.value });
  }

  handleError = () => {
    const { error } = this.props;
    if (error) {
      return error.message;
    }
    return '';
  }

  render() {
    const { username, password } = this.props.credentials
    const { submitting } = this.props;
    return (
      <div>
        <form className="bg-primary-light login-form" id="signup">
          <div className="color-primary to-center">
            <h2>LOG IN</h2>
          </div>
          <div className="form-errors">{this.handleError()}</div>
          <div className="input primary">
            <i className="fa fa-user" />
            <Input name="username" type="text" value={username} classes="bg-primary-light" placeholder="Username / Email" onChange={this.handleInput}/>
          </div>
          <div className="input primary">
            <i className="fa fa-lock" />
            <Input name="password" type="password" value={password} classes="bg-primary-light" placeholder="Password" onChange={this.handleInput}/>
          </div>
          <div className="align-right" id="forget-psswd">
            <a href="forgot-password-reset.html">Forget Password?</a>
          </div>
          <Button type="button" classes={`primary color-white ${
                submitting ? 'loading' : ''
              }`} color-white onClick={this.signin}>Log In</Button>
          <div className="icon-group">
            <div id="fb">
              <Link to="/#">
                <img src={fb} alt="fb-logo" />
              </Link>
            </div>
            <div id="twbs">
              <Link to="/#">
                <img src={twitter} alt="twbs-logo" />
              </Link>
            </div>
            <div id="gl">
              <Link to="/#">
                <img src={google} alt="gl-logo" />
              </Link>
            </div>
          </div>
          <div className="to-center" id="form-link">
            <span>Not a member?</span>
            <Link to="/#">Sign Up</Link>
          </div>
        </form>
      </div>
    );
  }
}
const mapStateToProps = ({ login : { error, credentials, submitting }}) => ({
  credentials,
  error,
  submitting
});
export default connect(mapStateToProps, { validateCredentials, loginUser, inputHandler })(LoginComponent);
