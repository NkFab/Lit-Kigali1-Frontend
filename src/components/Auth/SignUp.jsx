import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { isEmail } from 'validator';
import { handleSignUpForm, submitSignUp, socialAuth } from '../../redux/actions';
import Button from '../common/Button/Button';
import Input from '../common/Input/Input';
import fb from '../../assets/images/facebook.png';
import twitter from '../../assets/images/twitter.png';
import google from '../../assets/images/google.png';
import SocialLoginIcon from '../SocialLoginIcon';

export class SignUp extends Component {
  state = {
    validEmail: true,
    emailError: 'Invalid email',
    validPassword: true,
    passwordError: 'Minimum 6 characters',
    validUsername: true,
    usernameError: 'Required',
  };

  onSubmitButton = () => {
    const {
      onSubmit, email, username, password, history,
    } = this.props;
    this.setState({ validEmail: true, validPassword: true, validUsername: true });

    if (username.length < 1) {
      this.setState({
        validUsername: false,
        usernameError: 'Required',
      });
      return;
    }
    if (email.length < 1) {
      this.setState({
        validEmail: false,
        emailError: 'Required',
      });
      return;
    }
    if (!isEmail(email)) {
      this.setState({
        validEmail: false,
        emailError: 'Invalid email',
      });
      return;
    }

    if (password.length < 1) {
      this.setState({
        validPassword: false,
        passwordError: 'Required',
      });
      return;
    }
    if (password.length < 6) {
      this.setState({
        validPassword: false,
        passwordError: 'Minimum 6 characters',
      });
      return;
    }

    onSubmit({ email, password, username }).then((res) => {
      if (res.status === 201) {
        history.push('/home');
      }
    });
  };

  handleInputChange = (e) => {
    const { onInputChange } = this.props;
    onInputChange({ field: e.target.name, value: e.target.value });
  };

  renderErrors = () => {
    let { errors } = this.props;
    const { message } = this.props;
    errors = [message, ...errors];
    return errors.map(err => <span key={err}>{err.message || err}</span>);
  };

  socialAuthLogin = (provider) => {
    const { submitting, socialAuth } = this.props;

    if (!submitting) socialAuth(provider);
  };

  render() {
    const {
      username, password, email, submitting, onFlip,
    } = this.props;
    const {
      emailError,
      validEmail,
      validPassword,
      passwordError,
      validUsername,
      usernameError,
    } = this.state;
    return (
      <div>
        <form className="bg-primary-light login-form" id="signup">
          <div className="color-primary to-center">
            <h2>SIGN UP</h2>
          </div>
          <div className="form-errors" data-test="form-errors">
            {this.renderErrors()}
          </div>
          <div className="input primary">
            <i className="fa fa-user" />
            <Input
              name="username"
              type="text"
              placeholder="Username"
              classes="bg-primary-light"
              value={username}
              onChange={this.handleInputChange}
            />
            {validUsername ? '' : <div className="errors">{usernameError}</div>}
          </div>
          <div className="input primary">
            <i className="fa fa-at" />
            <Input
              name="email"
              type="email"
              placeholder="Email"
              classes="bg-primary-light"
              value={email}
              onChange={this.handleInputChange}
            />
            {validEmail ? '' : <div className="errors">{emailError}</div>}
          </div>
          <div className="input primary">
            <i className="fa fa-lock" />
            <Input
              name="password"
              type="password"
              placeholder="Password"
              classes="bg-primary-light"
              value={password}
              onChange={this.handleInputChange}
            />
            {validPassword ? '' : <div className="errors">{passwordError}</div>}
          </div>
          <Button
            classes={`primary color-white ${submitting ? 'loading' : ''}`}
            type="button"
            onClick={this.onSubmitButton}
          >
            Sign Up
          </Button>
          <div className="icon-group">
            <SocialLoginIcon
              id="fb"
              onClick={() => {
                this.socialAuthLogin('facebook');
              }}
              icon={fb}
              alt="fb-logo"
              dataTest={{ div: 'fb-container', a: 'fb-a', img: 'fb-img' }}
            />
            <SocialLoginIcon
              id="twbs"
              onClick={() => {
                this.socialAuthLogin('twitter');
              }}
              icon={twitter}
              alt="twbs-logo"
              dataTest={{ div: 'twbs-container', a: 'twbs-a', img: 'twbs-img' }}
            />
            <SocialLoginIcon
              id="gl"
              onClick={() => {
                this.socialAuthLogin('google');
              }}
              icon={google}
              alt="gl-logo"
              dataTest={{ div: 'gl-container', a: 'gl-a', img: 'gl-img' }}
            />
          </div>
          <div className="to-center" id="form-link">
            <span>Already a member?</span>
            <a href="#login" id="flip-signup" onClick={onFlip}>
              Sign In
            </a>
          </div>
        </form>
      </div>
    );
  }
}

export const mapStateToProps = state => ({
  username: state.signUp.username,
  email: state.signUp.email,
  password: state.signUp.password,
  message: state.signUp.message,
  errors: state.signUp.errors,
  submitting: state.signUp.submitting,
});

export const mapDispatchToProps = (dispatch, ownProps) => ({
  onInputChange: ({ field, value }) => dispatch(handleSignUpForm({ field, value })),
  onSubmit: ({ username, email, password }) => dispatch(
    submitSignUp({
      username,
      email,
      password,
      ownProps,
    }),
  ),
  socialAuth: provider => dispatch(socialAuth(provider)),
});

SignUp.propTypes = {
  email: PropTypes.string.isRequired,
  username: PropTypes.string.isRequired,
  password: PropTypes.string.isRequired,
  onSubmit: PropTypes.func.isRequired,
  message: PropTypes.string,
  onInputChange: PropTypes.func.isRequired,
  errors: PropTypes.array,
  submitting: PropTypes.bool,
  history: PropTypes.any.isRequired,
  socialAuth: PropTypes.func,
  onFlip: PropTypes.func,
};

SignUp.defaultProps = {
  message: '',
  errors: [],
  submitting: false,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(SignUp);
