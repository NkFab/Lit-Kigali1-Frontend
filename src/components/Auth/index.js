import React, { Component } from 'react';
import Login from './Login 2';

class AuthComponent extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  render() {
    return (
      <div className="main-content middle-content">
        <div id="card">
          <Login />
        </div>
      </div>
    );
  }
}

export default AuthComponent;
