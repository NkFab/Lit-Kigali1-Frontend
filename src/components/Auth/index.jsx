import React, { Component } from 'react';
import Signup from './SignUp';
import Login from './Login';

class LoginComponent extends Component {
  constructor(props) {
    super(props);
    this.card = React.createRef();
    this.sign = React.createRef();
    this.flip = this.flip.bind(this);
  }
  //   flipBack() {
  //     this.card.current.style.transform = 'rotate(180)deg';
  //   }

  //   flip() {
  //     this.setState({ card: flipBack });
  //   }
  flip() {
    this.card.current.style.transform = 'rotateY(180)deg';
  }

  render() {
    return (
      <div className="main-content middle-content">
        <button onClick={this.flip}>Flip that shit</button>
        <div id="card" ref={this.card}>
          <Signup />
          <Login />
        </div>
      </div>
    );
  }
}

export default LoginComponent;
