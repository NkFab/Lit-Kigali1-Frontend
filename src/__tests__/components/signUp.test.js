import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import { BrowserRouter as Router } from 'react-router-dom';
import { SignUp, mapStateToProps, mapDispatchToProps } from '../../components/Auth/SignUp';

let wrapper;
let store;
const mockFn = jest.fn();
const props = {
  email: '',
  password: '',
  username: '',
  onSubmit: jest.fn().mockImplementation(() => Promise.resolve({ status: 201 })),
  onInputChange: mockFn,
  history: { push: mockFn },
  socialAuth: mockFn,
  flip: mockFn,
};

const mockStore = configureMockStore();

const defaultState = {
  validUsername: true,
  usernameError: 'Required',
  validEmail: true,
  emailError: 'Invalid email',
  validPassword: true,
  passwordError: 'Minimum 6 characters',
};

describe('<SignUp />', () => {
  test('should render the <SignUp />', () => {
    const renderedValue = renderer
      .create(
        <Router>
          <SignUp {...props} />
        </Router>,
      )
      .toJSON();
    expect(renderedValue).toMatchSnapshot();
  });

  test('should render default state', () => {
    wrapper = shallow(<SignUp {...props} />);
    expect(wrapper.state()).toEqual(defaultState);
  });
});

describe('when clicking on submit button', () => {
  beforeEach(() => {
    wrapper = mount(
      <Router>
        <SignUp {...props} />
      </Router>,
    );
  });
  test('should call onSubmitFunction', () => {
    wrapper = shallow(<SignUp {...props} />);
    wrapper.find('Button').simulate('click');
    expect(wrapper.state()).toEqual({
      ...defaultState,
      validUsername: false,
    });
  });
  test('should set submitting to true', () => {
    wrapper = mount(<SignUp {...props} />);
    wrapper.setProps({ submitting: true });
    expect(wrapper.props().submitting).toBeTruthy();
    expect(wrapper.find('Button').hasClass('loading'));
  });
});

describe('when typing into the username input', () => {
  beforeAll(() => {
    wrapper = mount(<SignUp {...props} />);
    wrapper.find('Input[name="username"]').simulate('change', { target: { value: '' } });
  });
  test('should update the username prop', () => {
    wrapper = shallow(<SignUp {...props} />);
    expect(wrapper.state()).toEqual(defaultState);
  });
});

describe('when typing into the password input', () => {
  beforeAll(() => {
    wrapper = mount(<SignUp {...props} />);
    wrapper.find('Input[name="password"]').simulate('change', { target: { value: '' } });
  });
  test('should update the password prop', () => {
    wrapper = shallow(<SignUp {...props} />);
    expect(wrapper.state()).toEqual(defaultState);
  });
});

describe('when typing into the email input', () => {
  beforeAll(() => {
    wrapper = mount(<SignUp {...props} />);
    wrapper.find('Input[name="email"]').simulate('change', { target: { value: 'caleb@test.com' } });
  });
  test('should update the email prop', () => {
    wrapper = shallow(<SignUp {...props} />);
    expect(wrapper.state()).toEqual(defaultState);
  });
});

describe('when clicking on submit button', () => {
  beforeEach(() => {
    store = mockStore({});
    wrapper = mount(<SignUp store={store} {...props} />);
  });
  test('should call onSubmitFunction with wrong inputs', () => {
    wrapper.find('.button').simulate('click');
    expect(wrapper.state()).toEqual({
      ...defaultState,
      validUsername: false,
      passwordError: 'Minimum 6 characters',
    });
  });

  test('should call onSubmitFunction wrong email', () => {
    wrapper.setProps({ username: 'test', password: 'testpassword' });
    wrapper.find('.button').simulate('click');
    expect(wrapper.state()).toEqual({
      ...defaultState,
      validEmail: false,
      emailError: 'Required',
    });
  });

  test('should call onSubmitFunction wrong email', () => {
    wrapper.setProps({ username: 'test', password: 'testpassword', email: 'invalid@email' });
    wrapper.find('.button').simulate('click');
    expect(wrapper.state()).toEqual({
      ...defaultState,
      validEmail: false,
      emailError: 'Invalid email',
    });
  });

  test('should call onSubmitFunction wrong password', () => {
    wrapper.setProps({ username: 'test', email: 'test@email.com', password: '12345' });
    wrapper.find('.button').simulate('click');
    expect(wrapper.state()).toEqual({
      ...defaultState,
      validPassword: false,
      passwordError: 'Minimum 6 characters',
    });
  });

  test('should call onSubmitFunction wrong password', () => {
    wrapper.setProps({ username: 'test', email: 'test@email.com' });
    wrapper.find('.button').simulate('click');
    expect(wrapper.state()).toEqual({
      ...defaultState,
      validPassword: false,
      passwordError: 'Required',
    });
  });

  test('should call onSubmitFunction success', () => {
    wrapper.setProps({ username: 'test', email: 'test@email.com', password: 'testpassword' });
    wrapper.find('.button').simulate('click');
    expect(wrapper.state()).toEqual({
      ...defaultState,
    });
  });
});

describe('reducers', () => {
  test('should call onInputChange action', () => {
    const initialState = {
      signUp: {
        username: 'test',
        email: 'test@email.com',
        password: 'test123',
        message: 'message',
        errors: [],
        submitting: false,
      },
    };
    const state = mapStateToProps(initialState);
    expect(state).toEqual(initialState.signUp);
  });
});

describe('actions creators', () => {
  test('should call onInputChange action', () => {
    const dispatch = jest.fn();
    const payload = { field: 'email', value: 'email@email.com' };
    const expectedActions = {
      type: 'SIGNUP_FORM',
      payload,
    };
    mapDispatchToProps(dispatch).onInputChange(payload);
    expect(dispatch.mock.calls[0][0]).toEqual(expectedActions);
  });

  test('should call onSubmit action', () => {
    const dispatch = jest.fn();
    const payload = { username: 'test', email: 'test@email.com', password: 'testpassword' };
    mapDispatchToProps(dispatch).onSubmit(payload);
    expect(dispatch.mock.calls[0][0]).toBeDefined();
  });

  test('should call submit function', () => {
    wrapper.find('#flip').simulate('click');
    expect(props.flip).toHaveBeenCalled();
  });

  test('should call socialAuth action', () => {
    const dispatch = jest.fn();
    const payload = { provider: 'test' };
    mapDispatchToProps(dispatch).socialAuth(payload);
    expect(dispatch.mock.calls[0][0]).toBeDefined();
  });
});

describe('socialLogin', () => {
  test('should trigger click event', () => {
    const wrapper = mount(<SignUp store={store} {...props} />);
    wrapper
      .find('SocialLoginIcon')
      .at(0)
      .simulate('click');
    expect(props.socialAuth).toHaveBeenCalled();
  });
  test('should trigger click event', () => {
    const wrapper = mount(<SignUp store={store} {...props} />);
    wrapper
      .find('SocialLoginIcon')
      .at(1)
      .simulate('click');
    expect(props.socialAuth).toHaveBeenCalled();
  });
  test('should trigger click event', () => {
    const wrapper = mount(<SignUp store={store} {...props} />);
    wrapper
      .find('SocialLoginIcon')
      .at(2)
      .simulate('click');
    expect(props.socialAuth).toHaveBeenCalled();
  });
});
