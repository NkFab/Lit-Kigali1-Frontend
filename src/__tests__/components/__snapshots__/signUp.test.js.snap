// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp /> should render the <SignUp /> 1`] = `
<div
  className="main-content middle-content"
>
  <div
    id="card"
  >
    <form
      action="articles-create.html"
      className="bg-primary-light login-form"
      id="signup"
    >
      <div
        className="color-primary to-center"
      >
        <h2>
          SIGN UP
        </h2>
      </div>
      <div
        className="form-errors"
      >
        <span>
          
        </span>
      </div>
      <div
        className="input primary"
      >
        <i
          className="fa fa-user"
        />
        <input
          className="bg-primary-light"
          name="username"
          onChange={[Function]}
          placeholder="Username"
          type="text"
          value=""
        />
        
      </div>
      <div
        className="input primary"
      >
        <i
          className="fa fa-at"
        />
        <input
          className="bg-primary-light"
          name="email"
          onChange={[Function]}
          placeholder="Email"
          type="email"
          value=""
        />
        
      </div>
      <div
        className="input primary"
      >
        <i
          className="fa fa-lock"
        />
        <input
          className="bg-primary-light"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
      
      <button
        className="button primary color-white content-margin width-100 "
        disabled=""
        onClick={[Function]}
        type="button"
      >
        Sign Up
      </button>
      <div
        className="icon-group"
      >
        <div
          id="fb"
        >
          <img
            alt="fb-logo"
            src="test-file-stub"
          />
        </div>
        <div
          id="twbs"
        >
          <img
            alt="twbs-logo"
            src="test-file-stub"
          />
        </div>
        <div
          id="gl"
        >
          <img
            alt="gl-logo"
            src="test-file-stub"
          />
        </div>
      </div>
      <div
        className="to-center"
        id="form-link"
      >
        <span>
          Already a member?
        </span>
        <a
          to="/login"
        >
          Sign In
        </a>
      </div>
    </form>
  </div>
</div>
`;
